src/main/java/uk/gov/digital/ho/hocs/casework/api/DeadlineService.java

```diff
@@ -0,0 +1,99 @@
+package uk.gov.digital.ho.hocs.casework.api;
+
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import uk.gov.digital.ho.hocs.casework.api.dto.StageTypeDto;
+import uk.gov.digital.ho.hocs.casework.api.utils.DateUtils;
+import uk.gov.digital.ho.hocs.casework.client.infoclient.InfoClient;
+import uk.gov.digital.ho.hocs.casework.domain.model.BankHoliday;
+
+import java.time.LocalDate;
+import java.util.Comparator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * This service handles deadlines. These have been factored into their own service to avoid circular dependencies.
+ */
+@Slf4j
+@Service
+public class DeadlineService {
+
+    private final InfoClient infoClient;
+    private final BankHolidayService bankHolidayService;
+
+    @Autowired
+    public DeadlineService(final InfoClient infoClient,
+                           final BankHolidayService bankHolidayService) {
+        this.infoClient = infoClient;
+        this.bankHolidayService = bankHolidayService;
+    }
+
+    public LocalDate calculateWorkingDaysForStage(final String caseType,
+                                                  final LocalDate received,
+                                                  final LocalDate caseDeadline,
+                                                  final int workingDays) {
+        // -2 is a magic number denoting that the case deadline should be used for the stage, rather than calculating
+        // an individual deadline for the stage
+        if(workingDays == -2) {
+            return caseDeadline;
+        }
+
+        return calculateWorkingDaysForCaseType(caseType, received, workingDays);
+    }
+
+    public LocalDate calculateWorkingDaysForCaseType(String caseType, LocalDate receivedDate, int workingDays) {
+        final Set<LocalDate> bankHolidayDatesForCase = getBankHolidayDatesForCase(caseType);
+
+        return DateUtils.addWorkingDays(receivedDate, workingDays, bankHolidayDatesForCase);
+    }
+
+    public int calculateRemainingWorkingDaysForCaseType(String caseType, LocalDate receivedDate, LocalDate todaysDate) {
+        final Set<LocalDate> bankHolidayDatesForCase = getBankHolidayDatesForCase(caseType);
+
+        return DateUtils.calculateRemainingWorkingDays(todaysDate, receivedDate, bankHolidayDatesForCase);
+    }
+
+    /** todo: get final confirmation - should this method get the bank holidays from the Casework service,
```

Comment:
for me, the clue is if we're making lots of calls to info for data from only one place (with an exception of 1 call from workflow) then the data should reside locally to where we use it.